### ISSUE TITLE
How can I read the 'modified' flag?

### ISSUE THREAD

**Issue by Dr3ll** [immerjs/immer#471]
I am using the useImmerReducer hook.
When I inspect the state object that is given to my reducer function in the debugger I can see it has a flag 'modified' on its [[TARGET]] property. How can I read this flag in code?

**Comment by mweststrate** [comment]
It is not intended to be public api. What are you trying to achieved?

**Comment by Dr3ll** [comment]
In my app actions are called in the reducer functions of other actions, sometimes deferred as a callback to api requests.
A problem came up when I called an action that trigger a bunch of others from useEffect. All the actions were called but the state change from the first one was not correctly flushed to components.
My fix was to not dispatch actions directly but to queue them up when they are called in sequence; this queue is then advanced when a state change is flushed (so the change reflects immediately in components) or it is advanced when an action made no changes to the state object.
The annoying thing now is that in the reducer has to know that changes were made; with the modified flag I could automate this.

**Comment by mweststrate** [comment]
The problem sounds still a little bit too vague to me to justify a generic
change here, and calling actions from reducers is an anti pattern as far as
I can see . Beyond that checking modified state sounds like your are about
to loose the purity of your reducers (with the same inputs they might now
start to behave differently). I'd check if you can model this more
explicitly or check the Redux docs / SO how this problem in general is
addressed.

That all being said, you might be able to leverage `original(draft) ===
state` or similar checks.

**Comment by Dr3ll** [comment]
To my knowledge it only starts to be an antipattern when the dispatch of the next actions starts before the reducer is finished, which is not the case. The next action is simply scheduled to run after the one calling it.

**Comment by mweststrate** [comment]
sounds like the reducer has side effects? Anyway, not here to discuss Redux
patterns. I'll leave it open for a while, see if there is much need for it
by others, and then check if it not leads to easily to anti patterns.
Needing a modified concept sounds like you are keeping drafts around, which
is definitely an immer anti pattern. However, without seeing any line of
code this discussion remains pretty vague :)

**Comment by Dr3ll** [comment]
`private createReducer() {
       const [data, dispatcher] = useImmerReducer(this.reducer, this.initialState);
        this._dispatch = dispatcher;
        this._data = data;

	// If the state was changed this will be called again when the update is flushed. Then we can call the next action.
        this.advanceQueue(); 
}


private reducer(state: M, action: IAction<M>) {
        _this._lastComputedActionId = action.id;

        let patch = null;
        try {
            // The actions carry their own reducer function with them at it is called here (setProducts(), changeCategory(), fetchCatalog())
            // Nevermind how the actions are built
            patch = action.reduceFunc.apply(action._this, finalArgs);
        } catch (e) {
	    // error in reducer, bail out
            _this.advanceQueue();
            return state;
        }

        if (patch === undefined) {
            // The action function returned nothing, so I have to advance the queue. If I knew isModified here I could NOT advance it since it then will be done in createReducer()
            _this.advanceQueue();
            return state;
        }

	// We heavily assume at this point that the state was changed since the action function returned something.
	// We dont have to advance the queue since that will be done when the state updated is flushed.
	// If the state did NOT change we are fucked because the queue will not advance, we would need to know if it isModified
        return patch;
    }


private callAction(action: IAction<M>): void {
        // The action is placed in the stores queue
        const pre = this._actionQueue.length;

        logActionCall(action);

        this._actionQueue.push(action);

        // If the queue was empty there is no action currently underway, so just call this one
        if (pre === 0) {
            this._dispatch(action);
        }
}


private advanceQueue() {
        if (this._actionQueue.length > 0) {
            // Determine where the last called action is in the queue
            let deprecatedIndex = -1;
            for (let i = 0; i < this._actionQueue.length; i++) {
                if (this._actionQueue[i].id === this._lastComputedActionId) {
                    deprecatedIndex = i;
                    break;
                }
            }
            // Delete action that was called already
            if (deprecatedIndex >= 0) {
                this._actionQueue.splice(deprecatedIndex, 1);

                // Call next action from the top of the queue
                const [next] = this._actionQueue.slice(0);
                if (next) {
                    this._dispatch(next);
                }
            }
        }
}


// action
private setProducts(productPage: ICatalogPage, state?: IInternalCatalogState){
        state.products = productPage.data;

	return state;
}


// action
fetchCatalog(state?: IInternalCatalogState) {
        const {page, size, withVariants} = state.pagination;
	state.isLoading = true;

        const params = new CategoryParams(state.selectedCategory, page, size, withVariants);

	// We fetch the new product list and call the action that sets the products
        CatalogResource.catalog(params).then(
            (res) => {
                if (res.ok) {
                      Catalog().setProducts(res.body);
                }
            }
        );

	return state;
}


// action
changeCategory(category: string, state?: IInternalCatalogState) {
        state.selectedCategory = category;

	// When the category is changed we want to update the product list
        Catalog().fetchCatalog();

        return state;
}


someReactComponent = () => {
	
	const categories = [...];

	useEffect(() => {
        	Catalog().changeCategory(categories[0]);
	}, []);

	return ...
}`

This is a slightly simplified rundown of what is happening.
Maybe I am doing react wrong but it makes sense to me. I wouldn't know where else to call the catalog update in a sane way.

**Comment by mweststrate** [comment]
`isModified` is an internal implementation detail (depending on the type of data structure, and the proxy mode that is currently running, it might only be set only once the recipe finishes, and future changes might even cause it to have temporarily false positives). 

Closing this issue as I don't know how to contribute it without studying your code blob in depth for a significant amount of time. I recommend to check Stack overflow instead, as at this point the issue is not really actionable without playing consultant first :).

**Comment by Petersdavis** [comment]
Just an example use case for something like this:
1) we have a complex slice of state with many actions. 
2) we want to use marginal differences to communicate changes efficiently between server/client
3) for the marginal differences we implement a version flag that needs to increment when server/client actions modify the state. 

It would be convenient to be able to append:

 .addMatcher(
      (_action) => true, // This will match all actions
      (state, action) => {
       if(state.isModified){
          ++state.version
       }
      }
    );

**Comment by mweststrate** [comment]
you might try `current(draft) === original(draft)`, not sure it will be 100% accurate though (for example if you toggled a boolean twice), and it might be expensive for deep trees.

**Comment by Petersdavis** [comment]
I was thinking about wrapping the state in a Proxy upstream and unwrapping it to return the proxy target after processing.

**Comment by Petersdavis** [comment]
I'll give the current(draft) === original(draft)  and try something else if it becomes a performance issue checking on every action.  Thanks!
